internal/dataplane/parser/translators/ingress_atc_test.go

```diff
@@ -0,0 +1,259 @@
+						ResponseBuffering: kong.Bool(true),
+						RequestBuffering:  kong.Bool(true),
+						Tags:              kong.StringSlice("k8s-name:test-ingress", "k8s-namespace:default"),
+					},
+					ExpressionRoutes: true,
+				}},
+				Backends: []kongstate.ServiceBackend{{
+					Name:      "test-service",
+					Namespace: corev1.NamespaceDefault,
+					PortDef: kongstate.PortDef{
+						Mode:   kongstate.PortModeByNumber,
+						Number: 80,
+					},
+				}},
+				Parent: expectedParentIngress(),
+			}},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			services := TranslateIngress(tc.ingress, false, true)
+			checkOnlyObjectMeta := cmp.Transformer("checkOnlyObjectMeta", func(i *netv1.Ingress) *netv1.Ingress {
+				// In the result we only care about ingresses' metadata being equal.
+				// We ignore specification to simplify tests.
+				return &netv1.Ingress{
+					ObjectMeta: i.ObjectMeta,
+				}
+			})
+			diff := cmp.Diff(tc.expectedServices, services, checkOnlyObjectMeta)
+			require.Empty(t, diff, "expected no difference between expected and translated ingress")
+		})
+	}
+}
+
+func TestPathMatcherFromIngressPath(t *testing.T) {
+	testCases := []struct {
+		name        string
+		path        netv1.HTTPIngressPath
+		regexPrefix string
+		expression  string
+	}{
+		{
+			name: "simple prefix match",
+			path: netv1.HTTPIngressPath{
+				Path:     "/v1/api",
+				PathType: &pathTypePrefix,
+			},
+			expression: `(http.path == "/v1/api") || (http.path ^= "/v1/api/")`,
+		},
+		{
+			name: "simple exact match",
+			path: netv1.HTTPIngressPath{
+				Path:     "/v1/api",
+				PathType: &pathTypeExact,
+			},
+			expression: `http.path == "/v1/api"`,
+		},
+		{
+			name: "regex match",
+			path: netv1.HTTPIngressPath{
+				Path:     "/~/[a-z]+",
+				PathType: &pathTypeImplementationSpecific,
+			},
+			expression: `http.path ~ "^/[a-z]+"`,
+		},
+		{
+			name: "empty path",
+			path: netv1.HTTPIngressPath{
+				Path:     "",
+				PathType: &pathTypePrefix,
+			},
+			expression: `http.path ^= "/"`,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			regexPrefix := tc.regexPrefix
+			if regexPrefix == "" {
+				regexPrefix = ControllerPathRegexPrefix
+			}
+			matcher := pathMatcherFromIngressPath(tc.path, regexPrefix)
+			require.Equal(t, tc.expression, matcher.Expression())
+		})
+	}
+}
+
+func TestHeaderMatcherFromHeaders(t *testing.T) {
+	testCases := []struct {
+		name       string
+		headers    map[string][]string
+		expression string
+	}{
+		{
+			name: "single header with single value",
+			headers: map[string][]string{
+				"X-Key1": {"value1"},
+			},
+			expression: `http.headers.x_key1 == "value1"`,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			matcher := headerMatcherFromHeaders(tc.headers)
+			require.Equal(t, tc.expression, matcher.Expression())
+		})
+	}
+}
```

Comment:
I think we're still missing UTs for below functions, would you mind adding them?
- `hostMatcherFromIngressHosts`
- `protocolMatcherFromProtocols`
- `methodMatcherFromMethods`
- `sniMatcherFromSNIs`.

Reply:
Added tests for them.